### ISSUE TITLE
Window crashes if starting in the 2nd display (or move to the 2nd display)

### ISSUE THREAD

**Issue by greasycat** [neovide/neovide#1342]
**Describe the bug**

**To Reproduce**
1. start neovide
2. drag the window and move to a different display
3. texts disappear
4. click or press any key on keyboard, the window crashes

**Expected behavior**
Normal render and no crash

**Screenshots**

![Steps](https://i.ibb.co/B38MPrn/ezgif-com-gif-maker.gif)

**Desktop (please complete the following information):**
 - OS:
 ```
                   -`                    rf@archlinux 
                  .o+`                   ------------ 
                 `ooo/                   OS: Arch Linux x86_64 
                `+oooo:                  Kernel: 5.18.1-arch1-1 
               `+oooooo:                 Uptime: 26 mins 
               -+oooooo+:                Packages: 1588 (pacman), 10 (flatpak) 
             `/:-:++oooo+:               Shell: bash 5.1.16 
            `/++++/+++++++:              Resolution: 2560x1440, 3440x1440 
           `/++++++++++++++:             DE: Plasma 5.24.5 
          `/+++ooooooooooooo/`           WM: KWin 
         ./ooosssso++osssssso+`          WM Theme: Arc-Dark 
        .oossssso-````/ossssss+`         Theme: [Plasma], Breeze [GTK2/3] 
       -osssssso.      :ssssssso.        Icons: breeze-dark [Plasma], breeze-dark [GTK2/3] 
      :osssssss/        osssso+++.       Terminal: konsole 
     /ossssssss/        +ssssooo/-       CPU: AMD Ryzen 9 5900X (24) @ 3.700GHz 
   `/ossssso+/:-        -:/+osssso+-     GPU: NVIDIA GeForce RTX 3090 
  `+sso+:-`                 `.-/+oso:    Memory: 3554MiB / 32070MiB 
 `++:.                           `-/+/
```
 - Neovide Version 0 8
 - Neovim Version 0.7.0

```
thread 'main' panicked at 'Could not create surface', src/renderer/rendered_window.rs:75:6
stack backtrace:
   0:     0x555a2a59ae8c - std::backtrace_rs::backtrace::libunwind::trace::h3d7cd6069758c80f
                               at /rustc/1.61.0/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x555a2a59ae8c - std::backtrace_rs::backtrace::trace_unsynchronized::h89a22958833bc026
                               at /rustc/1.61.0/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x555a2a59ae8c - std::sys_common::backtrace::_print_fmt::h185896f4f82c8242
                               at /rustc/1.61.0/library/std/src/sys_common/backtrace.rs:66:5
   3:     0x555a2a59ae8c - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hbd89b9cf25956154
                               at /rustc/1.61.0/library/std/src/sys_common/backtrace.rs:45:22
   4:     0x555a2a5c220c - core::fmt::write::h526dc76e7de39ee8
                               at /rustc/1.61.0/library/core/src/fmt/mod.rs:1194:17
   5:     0x555a2a594d11 - std::io::Write::write_fmt::hdb3a6d87b62a7672
                               at /rustc/1.61.0/library/std/src/io/mod.rs:1655:15
   6:     0x555a2a59c8f5 - std::sys_common::backtrace::_print::h33b2931cb6ebf7d3
                               at /rustc/1.61.0/library/std/src/sys_common/backtrace.rs:48:5
   7:     0x555a2a59c8f5 - std::sys_common::backtrace::print::h2fefd24a46d56ae3
                               at /rustc/1.61.0/library/std/src/sys_common/backtrace.rs:35:9
   8:     0x555a2a59c8f5 - std::panicking::default_hook::{{closure}}::hb52dd86f3e15f1d5
                               at /rustc/1.61.0/library/std/src/panicking.rs:295:22
   9:     0x555a2a59c598 - std::panicking::default_hook::h15c0d734222be4c0
                               at /rustc/1.61.0/library/std/src/panicking.rs:314:9
  10:     0x555a2a59ce53 - std::panicking::rust_panic_with_hook::h8ef8647bb98a4b13
                               at /rustc/1.61.0/library/std/src/panicking.rs:698:17
  11:     0x555a2a59cd37 - std::panicking::begin_panic_handler::{{closure}}::hecadb3dd74c77976
                               at /rustc/1.61.0/library/std/src/panicking.rs:588:13
  12:     0x555a2a59b324 - std::sys_common::backtrace::__rust_end_short_backtrace::ha03a690be8d1d9a2
                               at /rustc/1.61.0/library/std/src/sys_common/backtrace.rs:138:18
  13:     0x555a2a59ca59 - rust_begin_unwind
                               at /rustc/1.61.0/library/std/src/panicking.rs:584:5
  14:     0x555a29558a33 - core::panicking::panic_fmt::h5349b352d3b6f3f7
                               at /rustc/1.61.0/library/core/src/panicking.rs:143:14
  15:     0x555a2a5c0161 - core::panicking::panic_display::hce3d3b74feb2abfd
                               at /rustc/1.61.0/library/core/src/panicking.rs:72:5
  16:     0x555a2a5c010b - core::panicking::panic_str::h22a5a95120322b75
                               at /rustc/1.61.0/library/core/src/panicking.rs:56:5
  17:     0x555a295588a6 - core::option::expect_failed::hed605dae359c1913
                               at /rustc/1.61.0/library/core/src/option.rs:1874:5
  18:     0x555a2977a3ae - core::option::Option<T>::expect::hc540dd8b250ea5c7
                               at /rustc/1.61.0/library/core/src/option.rs:718:21
  19:     0x555a297169be - neovide::renderer::rendered_window::build_window_surface::he2fad4b6e051cd8e
                               at /home/rf/Build/neovide/src/renderer/rendered_window.rs:66:5
  20:     0x555a29716aa5 - neovide::renderer::rendered_window::build_window_surface_with_grid_size::h58fa0833c2e6add8
                               at /home/rf/Build/neovide/src/renderer/rendered_window.rs:83:23
  21:     0x555a2971801f - neovide::renderer::rendered_window::RenderedWindow::handle_window_draw_command::ha4edffb4be6564d3
                               at /home/rf/Build/neovide/src/renderer/rendered_window.rs:466:48
  22:     0x555a295f8aff - neovide::renderer::Renderer::handle_draw_command::h277a27cab4763364
                               at /home/rf/Build/neovide/src/renderer/mod.rs:209:25
  23:     0x555a295f79b4 - neovide::renderer::Renderer::draw_frame::h6c1bac316ffb0c16
                               at /home/rf/Build/neovide/src/renderer/mod.rs:132:13
  24:     0x555a295d2824 - neovide::window::GlutinWindowWrapper::draw_frame::h25ef2e0f3a19f68d
                               at /home/rf/Build/neovide/src/window/mod.rs:191:28
  25:     0x555a296a201f - neovide::window::create_window::{{closure}}::hff5b2ce56d5c86a6
                               at /home/rf/Build/neovide/src/window/mod.rs:428:13
  26:     0x555a2970b162 - winit::platform_impl::platform::sticky_exit_callback::hf59ad695d356bca9
                               at /home/rf/.cargo/git/checkouts/winit-58efae9df3f5264a/59eb01c/src/platform_impl/linux/mod.rs:760:5
  27:     0x555a295878f7 - winit::platform_impl::platform::x11::EventLoop<T>::run_return::h04616173b269bda8
                               at /home/rf/.cargo/git/checkouts/winit-58efae9df3f5264a/59eb01c/src/platform_impl/linux/x11/mod.rs:303:13
  28:     0x555a29588c83 - winit::platform_impl::platform::x11::EventLoop<T>::run::h4025a8dcac744883
                               at /home/rf/.cargo/git/checkouts/winit-58efae9df3f5264a/59eb01c/src/platform_impl/linux/x11/mod.rs:425:9
  29:     0x555a2970ae96 - winit::platform_impl::platform::EventLoop<T>::run::h5e13089727ec49fa
                               at /home/rf/.cargo/git/checkouts/winit-58efae9df3f5264a/59eb01c/src/platform_impl/linux/mod.rs:676:56
  30:     0x555a29719fba - winit::event_loop::EventLoop<T>::run::h079b9d3df55c457c
                               at /home/rf/.cargo/git/checkouts/winit-58efae9df3f5264a/59eb01c/src/event_loop.rs:154:9
  31:     0x555a295d49e5 - neovide::window::create_window::hea3697a018bd904c
                               at /home/rf/Build/neovide/src/window/mod.rs:404:5
  32:     0x555a2963958e - neovide::main::he2cba057125a0fcf
                               at /home/rf/Build/neovide/src/main.rs:144:5
  33:     0x555a2966348b - core::ops::function::FnOnce::call_once::h9e059957da20b4eb
                               at /rustc/1.61.0/library/core/src/ops/function.rs:227:5
  34:     0x555a2957747e - std::sys_common::backtrace::__rust_begin_short_backtrace::hc89d4c1b334bf691
                               at /rustc/1.61.0/library/std/src/sys_common/backtrace.rs:122:18
  35:     0x555a297842a1 - std::rt::lang_start::{{closure}}::h99a259ebadee3010
                               at /rustc/1.61.0/library/std/src/rt.rs:145:18
  36:     0x555a2a58cb6e - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::hff1d360addf0e0a5
                               at /rustc/1.61.0/library/core/src/ops/function.rs:259:13
  37:     0x555a2a58cb6e - std::panicking::try::do_call::haaf1332f62982d76
                               at /rustc/1.61.0/library/std/src/panicking.rs:492:40
  38:     0x555a2a58cb6e - std::panicking::try::hb1a5ffffc51ce834
                               at /rustc/1.61.0/library/std/src/panicking.rs:456:19
  39:     0x555a2a58cb6e - std::panic::catch_unwind::h8b3ba16b791f0a1a
                               at /rustc/1.61.0/library/std/src/panic.rs:137:14
  40:     0x555a2a58cb6e - std::rt::lang_start_internal::{{closure}}::hc3bf7d68bdc46579
                               at /rustc/1.61.0/library/std/src/rt.rs:128:48
  41:     0x555a2a58cb6e - std::panicking::try::do_call::hb9549501a638587d
                               at /rustc/1.61.0/library/std/src/panicking.rs:492:40
  42:     0x555a2a58cb6e - std::panicking::try::h56ce4741a577b244
                               at /rustc/1.61.0/library/std/src/panicking.rs:456:19
  43:     0x555a2a58cb6e - std::panic::catch_unwind::he844618d24ada429
                               at /rustc/1.61.0/library/std/src/panic.rs:137:14
  44:     0x555a2a58cb6e - std::rt::lang_start_internal::h63ec520e87e83a8d
                               at /rustc/1.61.0/library/std/src/rt.rs:128:20
  45:     0x555a29784270 - std::rt::lang_start::h145e7c6371d77d7c
                               at /rustc/1.61.0/library/std/src/rt.rs:144:17
  46:     0x555a2963996c - main
  47:     0x7f5d14a29290 - <unknown>
  48:     0x7f5d14a2934a - __libc_start_main
  49:     0x555a29559005 - _start
                               at /build/glibc/src/glibc/csu/../sysdeps/x86_64/start.S:115
  50:                0x0 - <unknown>

```

**Additional context**
```
server glx version string: 1.4
client glx version string: 1.4
GLX version: 1.4
OpenGL core profile version string: 4.6.0 NVIDIA 515.43.04
OpenGL core profile shading language version string: 4.60 NVIDIA
OpenGL version string: 4.6.0 NVIDIA 515.43.04
OpenGL shading language version string: 4.60 NVIDIA
OpenGL ES profile version string: OpenGL ES 3.2 NVIDIA 515.43.04
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.20
    GL_EXT_shader_group_vote, GL_EXT_shader_implicit_conversions, 
```

**Comment by greasycat** [comment]
Update: the window cannot be created in the second display

**Comment by MultisampledNight** [comment]
Hm, are you running Plasma w/ Wayland by chance?

**Comment by greasycat** [comment]
> Hm, are you running Plasma w/ Wayland by chance?

No, It's X11

**Comment by codelerk** [comment]
Mine does this error when just running the program on single monitor laptop. It crashes and gives me this error. I am not sure why it is doing this. I have tried uninstalling and recompiling and nothing seems to work to fix this issue.

**Comment by MultisampledNight** [comment]
I assume it's yet another issue with over/underflowing coordinates, but I haven't been able to reproduce this for now. A debug build running w/ `--nofork` would probably panic at an interesting location.

**Comment by codelerk** [comment]
thread 'main' panicked at 'Could not create surface', src/renderer/rendered_window.rs:75:6
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

This is what i get when trying to run neovide.

**Comment by MultisampledNight** [comment]
Yes, it's the same as above. Though I meant running Neovide in debug mode, in other words from source without the `--release` flag passed to `cargo run`.

**Comment by MultisampledNight** [comment]
Cargo is Rust's build system, yep. You can get the toolchain from https://rustup.rs/.

**Comment by codelerk** [comment]
does the --release need to be changed to --debug? My host system is arch linux.  or does it need to just be cargo build I found it in the PKGBUILD file under the build function?

**Comment by MultisampledNight** [comment]
No, just leave the flag away.

**Comment by codelerk** [comment]
yeah just gives me the same error nothing else added.

**Comment by codelerk** [comment]
Finished dev [unoptimized + debuginfo] target(s) in 0.10s
     Running `target/debug/neovide --nofork`
thread 'main' panicked at 'Could not create surface', src/renderer/rendered_window.rs:75:6
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

**Comment by MultisampledNight** [comment]
Nah, no worries, you did nothing wrong. It was just an assumption anyways.

**Comment by greasycat** [comment]
After checking a few values, it turned out my problem was caused by the `scale_factor` update from winit. In simpler terms, winit failed to calculate the right scale factor of X11.

The scale factor in my first working display is 1.16, even though it should be 1.0 based on my X11 setting.
After moving to the 2nd display, the scale factor peaked at 588.833

# Temporaily fix
As @dhardy suggested, Forcibly setting the scale factor temporarily solves the issue
```bash
export WINIT_X11_SCALE_FACTOR=1
```
https://github.com/rust-windowing/winit/issues/2229
https://github.com/rust-windowing/winit/issues/2231
https://github.com/rust-windowing/winit/issues/1983

# Or update winit
The issue should be solved based on the following code

https://github.com/rust-windowing/winit/blob/57897780371c621be39e4c18f89aecf92ce02ae1/src/platform_impl/linux/x11/util/randr.rs#L16-L35

**Comment by MultisampledNight** [comment]
I've backported https://github.com/rust-windowing/winit/commit/b4774861db2f765ff016b1a3370ebf03debbd30d to our winit fork, @greasycat could you try out whether rebuilding from main removes the need for your workaround?

**Comment by greasycat** [comment]
@MultisampledNight Problem solved, can you update Cargo.lock as well?

**Comment by MultisampledNight** [comment]
Done, GL HF.
